Add troubleshooting document (#2700)
Azure/typespec-azure: website/src/content/docs/docs/migrate-swagger/faq/pipeline.md

2. Check the report on the "Summary" page
3. The output should match exactly what you see in [this step](../01-get-started.md#review-and-adjust-the-typespec) on the local machine
4. Review the changes to verify they are expected

### Single Swagger File Before Migration

If you have only one Swagger file in the latest version, use this pipeline to detect breaking changes. If it fails, refer to [Resolving Swagger Breaking Change Violations](./faq/breakingchange.md).

**Known Issues**: The following pipeline failures are false alerts and can be safely ignored:

#### 1017 - ReferenceRedirection

This typically occurs when an inlined anonymous enum becomes a named enum.

**Before:**

```json
"SomeModel": {
  "enumProperty": {
    "type": "string",
    "enum": [
      "If",
      "Else",
      "None"
    ],
    "x-ms-enum": {
      "name": "EnumType",
      "modelAsString": true
    }
  }
}
```

**After:**

```json
"SomeModel": {
  "enumProperty": {
    "$ref": "#/definitions/EnumType"
  }
},
"EnumType": {
  "type": "string",
  "enum": [
    "If",
    "Else",
    "None"
  ],
  "x-ms-enum": {
    "name": "EnumType",
    "modelAsString": true
  }
}
```

#### 1023 - TypeFormatChanged

> The error message `The new version has a different format 'uri' than the previous one ''.` on the `nextLink` property is expected because the [page template defines the `nextLink` as `uri`](./mustread.md#using-page-model-from-azurecore-library).

Author: The error message does have `''`.

> The error message `The new version has a different format 'uri' than the previous one ''.` on the `nextLink` property is expected because the [page template defines the `nextLink` as `uri`](./mustread.md#using-page-model-from-azurecore-library).

Author: meaning the previous format is empty.

**Before:**

```json
"SomeModel": {
  "someProperty": {
    "type": "string",
    "enum": [
      "If"
    ],
    "x-ms-enum": {
      "name": "someName",
      "modelAsString": true
    }
  }
}
```

**After:**

```json
"SomeModel": {
  "someProperty": {
    "$ref": "#/definitions/SomeName"
  }
},
"SomeName": {
  "type": "string",
  "enum": [
    "If"
  ],
  "x-ms-enum": {
    "name": "SomeName",
    "modelAsString": true
  }
}
```

## Swagger ModelValidation

Some validation rules may fail because TypeSpec fixes certain legacy patterns. Review the known issues below and take appropriate action.

### INVALID_FORMAT: Object didn't pass validation for format uri: {nextlink}

#### Root Cause

The response model of a pageable operation uses the Azure.Core.Page template:

```typespec
model ResponseModel is Azure.Core.Page<ItemType>;
```

This standardized approach changes the `nextLink` type from plain `string` to `uri`. If your previous example's `nextLink` value isn't a valid URI, it will cause a validation error.

#### Resolution

Choose one of the following options:

1. **Update your example** to include a valid URI value, or

> 2. **Modify your TypeSpec response model** to use a custom definition:

Author: It makes more sense to change it to: 
2. **Modify your TypeSpec response model** to use a custom definition if the nextLink returned by the service is not a URI:

---
title: Resolving Pipeline failures
---

> This document explains how to mitigate pipeline failures in TypeSpec migration PRs.

Reviewer: There is already a document in the spec repo for this, we should at the very least link it as it contains more info

Reviewer: we might also want to reorganize a bit here as this is at the same level and after the resolving "Breaking change failues" and resolving lint diff failures

Author: > There is already a document in the spec repo for this, we should at the very least link it as it contains more info

Which document you are saying? The content in this document only applies to migration cases.

---
title: Resolving Pipeline failures
---

This document explains how to mitigate pipeline failures in TypeSpec migration PRs.

## Swagger BreakingChange

> This pipeline will definitely fail if you have more than one swagger files in your latest version. See detailed explanation in the description of this [issue](https://github.com/Azure/typespec-azure/issues/2194#issue-2844564216). Therefore, we leverage "TypeSpec Migration Validation" pipeline to give us a signal if there is breaking change. Go to this pipeline and check the output from "Run TypeSpec Migration Validation". The output is expected to be exactly the same as this [step](../01-get-started.md#review-and-adjust-the-typespec) in your local machine. Review the changes to see if it's expected.

Reviewer 2: ```suggestion
This pipeline will fail if you have more than one swagger file in your latest version. See detailed explanation in the description of this [issue](https://github.com/Azure/typespec-azure/issues/2194#issue-2844564216). Therefore, we leverage the "TypeSpec Migration Validation" pipeline to properly identify real breaking changes. Go to this pipeline and check the output from "Run TypeSpec Migration Validation". The output is expected to be exactly the same as this [step](../01-get-started.md#review-and-adjust-the-typespec) in your local machine. Review the changes to see if they are expected.
```

> However, if you have only one swagger file in your latest version, you should check out the results of this pipeline. Refer to [Resolving Swagger Breaking Change Violations](./faq/breakingchange.md) if it fails.

Reviewer 2: ```suggestion
However, if you have only one swagger file in your latest version, you should use this pipeline to detect breaking changes. Refer to [Resolving Swagger Breaking Change Violations](./faq/breakingchange.md) if it fails.
```

## Swagger ModelValidation

Some validation rules may fail because TypeSpec fixes certain legacy patterns. Review the known issues below and take appropriate action.

### INVALID_FORMAT: Object didn't pass validation for format uri: {nextlink}

#### Root cause

The response model of a pageable operation uses the Azure.Core.Page template:

```typespec
model ResponseModel is Azure.Core.Page<ItemType>;
```

This standardized approach makes the nextLink type `uri` instead of plain `string`. If your previous example's `nextLink` value isn't a valid URI, it will cause a CI error.

#### Mitigation

Either:

1. Update your example to include a valid URI value, or
2. Modify your TypeSpec response model definition to:

```typespec
@pagedResult
model ResponseModel {
  @items
  value: ItemType[];

  @nextLink
  nextLink?: string;
}
```

### INVALID_FORMAT: Object didn't pass validation for format arm-id

#### Root cause

> It is possibly because your own defined resource is mapped to the resource defined in common-type. See [this](./breakingchange.md#using-resources-from-common-types) for details. The `id` type for `Resource` in common-types is of format `arm-id`.

Reviewer 2: This is most likely because they didn't select the appropriate common-types version  (versions v5 and later use arm-id consistently).  We should suggest that they use the Azure.Core scalar `armResourceIdentifier` for these types (or the equivalent scalar aliased in Azure.ResourceManager.
